daemon/src/rollover_taker.rs

```diff
@@ -0,0 +1,318 @@
+use crate::address_map::Stopping;
+use crate::connection;
+use crate::model::cfd::{Cfd, Dlc, OrderId, Role, RollOverProposal, SettlementKind};
+use crate::model::{BitMexPriceEventId, Timestamp};
+use crate::oracle::{self, GetAnnouncement};
+use crate::projection::{self, UpdateRollOverProposal};
+use crate::setup_contract::{self, RolloverParams};
+use crate::tokio_ext::spawn_fallible;
+use crate::wire::{self, RollOverMsg};
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use futures::channel::mpsc::{self, UnboundedSender};
+use futures::{future, SinkExt};
+use maia::secp256k1_zkp::schnorrsig;
+use xtra::prelude::MessageChannel;
+use xtra_productivity::xtra_productivity;
+
+pub struct Actor {
+    cfd: Cfd,
+    n_payouts: usize,
+    oracle_pk: schnorrsig::PublicKey,
+    timestamp: Timestamp,
+    maker: xtra::Address<connection::Actor>,
+    get_announcement: Box<dyn MessageChannel<GetAnnouncement>>,
+    projection: xtra::Address<projection::Actor>,
+    on_completed: Box<dyn MessageChannel<Completed>>,
+    on_stopping: Vec<Box<dyn MessageChannel<Stopping<Self>>>>,
+    rollover_msg_sender: Option<UnboundedSender<RollOverMsg>>,
+}
+
+impl Actor {
+    pub fn new(
+        (cfd, n_payouts): (Cfd, usize),
+        oracle_pk: schnorrsig::PublicKey,
+        maker: xtra::Address<connection::Actor>,
+        get_announcement: &(impl MessageChannel<GetAnnouncement> + 'static),
+        projection: xtra::Address<projection::Actor>,
+        on_completed: &(impl MessageChannel<Completed> + 'static),
+        (on_stopping0, on_stopping1): (
+            &(impl MessageChannel<Stopping<Self>> + 'static),
+            &(impl MessageChannel<Stopping<Self>> + 'static),
+        ),
+    ) -> Self {
+        Self {
+            cfd,
+            n_payouts,
+            oracle_pk,
+            timestamp: Timestamp::now(),
+            maker,
+            get_announcement: get_announcement.clone_channel(),
+            projection,
+            on_completed: on_completed.clone_channel(),
+            on_stopping: vec![on_stopping0.clone_channel(), on_stopping1.clone_channel()],
+            rollover_msg_sender: None,
+        }
+    }
+
+    async fn propose(&self, this: xtra::Address<Self>) -> Result<()> {
+        self.maker
+            .send(connection::ProposeRollOver {
+                order_id: self.cfd.order.id,
+                timestamp: self.timestamp,
+                address: this,
+            })
+            .await??;
+
+        self.update_proposal(Some((
+            RollOverProposal {
+                order_id: self.cfd.order.id,
```

Comment:
This seems redundant? We already include the ID in the `UpdateRollOverProposal` struct.

Reply:
Do you mean that you think we should remove the `order` from `RollOverProposal`. I don't think I added this stuff.

Reply:
It seems that we can optimize this a bit. I also think this timestamp is not actually used anywhere :rofl: 

Don't need to change it now!